src/prefect/agent/kubernetes/agent.py

```diff
@@ -103,9 +104,153 @@ def __init__(
             config.load_kube_config()
 
         self.batch_client = client.BatchV1Api()
+        self.core_client = client.CoreV1Api()
+        self.k8s_client = client
 
         self.logger.debug(f"Namespace: {self.namespace}")
 
+    def _get_prefect_jobs(self) -> list:
+        """
+        Get a list of prefect jobs that are currently present in the cluster.
+
+        Returns:
+            - A list of dictionaries where each entry matches a job name, its pod names, and flow run id
+        """
+
+        jobs_list = list()
+
+        self.logger.debug(
+            "Retrieving information of jobs that are currently in the cluster..."
+        )
+
+        more = True
+        _continue = ""
+        while more:
+            try:
+                jobs = self.batch_client.list_namespaced_job(
+                    namespace=self.namespace,
+                    label_selector="prefect.io/identifier",
+                    limit=20,
+                    _continue=_continue,
+                )
+                _continue = jobs.metadata._continue
+                more = bool(_continue)
+
+                for job in jobs.items:
+                    pods = self.core_client.list_namespaced_pod(
+                        namespace=self.namespace,
+                        label_selector="prefect.io/identifier={}".format(
+                            job.metadata.labels.get("prefect.io/identifier")
+                        ),
+                    )
+
+                    pod_names = []
+                    for pod in pods.items:
+                        pod_names.append(pod.metadata.name)
+
+                    jobs_list.append(
+                        {
+                            "job_name": job.metadata.name,
+                            "pod_names": pod_names,
+                            "flow_run_id": job.metadata.labels.get(
+                                "prefect.io/flow_run_id"
+                            ),
+                        }
+                    )
+            except self.k8s_client.rest.ApiException as exc:
+                if exc.status == 410:
+                    self.logger.debug("Refreshing job listing token...")
+                    _continue = ""
+                    continue
+
+        self.logger.debug(
+            f"Found {len(jobs_list)} existing prefect jobs in the cluster"
+        )
+        return jobs_list
+
+    def _read_prefect_jobs(self, jobs: list) -> None:
```

Comment:
This method doesn't really "read prefect jobs" (since it might delete finished/failed jobs, and also notifies the cloud api). Could you rename to a more descriptive name?

Reply:
Since one method is listing jobs, and the other is re-iterating through those jobs to respond, perhaps this should really be one method, and jobs are handled individually as they're found in the pagination loop? The separation here feels a bit off.

Reply:
Thinking more (not to tell you how to structure your code). Perhaps this would be cleaner as:

- One method that yields tuples of job + pods, performing the pagination as it goes (a generator method). This would prevent loading too many jobs/pods into memory at once.
- One method for handling a single job + pods tuple, doing whatever actions are required.
- One method for doing the whole loop over jobs/pods + processing

Seems simpler to structure and test, as the generator method could be mocked out, and the actions we take on the retrieved info are the tricky bits.